onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -86,6 +91,40 @@ Status SliceOutUnwantedOutputSection(cudaStream_t stream,
   return SliceCuda::Impl(stream, input_data, input_dims, output_data, compute_metadata, element_size);
 }
 
+template <typename T, bool NHWC>
+Status Conv<T, NHWC>::PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
+                bool& is_packed, [[maybe_unused]] PrePackedWeights* prepacked_weights) {
+  is_packed = false;
+  // only layout of weight input is adjusted via PrePack
+  if (NHWC) {  // InputTensors::IN_W
+    if (input_idx == 1) {
+      // Transpose from {M, C/group, kH, kW} to {M, kH, kW, C/group}
+      auto orig_shape = tensor.Shape();
+
+      InlinedVector<size_t> perm{0, 2, 3, 1};
+      gsl::span<size_t> permutation(perm.data(), 4);
+      TensorShapeVector new_dims{orig_shape[0],
+                                orig_shape[2],
+                                orig_shape[3],
+                                orig_shape[1]};
+      W_ = Tensor::Create(tensor.DataType(), TensorShape(new_dims), std::move(alloc));
+
+      cudaStream_t stream = 0;
+      auto status = cuda::Transpose::DoTranspose(GetDeviceProp(),
+                                  stream,
```

Comment:
Thinking out aloud - Is there any value in doing a `cudaStreamSync()` before we exit `PrePack()` ? `PrePack()` is only invoked at model loading time and hence the performance implication with using a sync is negligible. Without this, there is no guarantee that pre-packing of the weights would have completed when we return back from `PrePack()`. Any refactoring if any in the calling layer will require knowledge of this information (that pre-packing may not have completed)

Reply:
Actually good that you mention this ! I wanted to look into how to acquire the compute stream of the session in pre pack, but did not manage to when I tried so simply used stream 0 here. I will try to look into getting the compute stream - happy about any recommendations.

Reply:
It would be nice to use the compute stream dedicated for the session to use to perform the prepacking operation. Unless it requires wholesale rework of some component, that approach would be preferable than using the default stream. In any case, I think it will be better to sync and ensure that the perpacking op has gone through successfully before returning, otherwise the calling layer needs to have the logic to ensure that prepacking was successful before accepting `Run()` requests.

Reply:
I switched to using the providers compute stream. Could you guide me to when that sync would have to be done to sync after all prepacks ?

Reply:
Sync was added in `PrePack()`